plane/src/typed_unix_socket/client.rs

```diff
@@ -0,0 +1,247 @@
+{
+    tx: broadcast::Sender<WrappedMessage<MessageToServer>>,
+    response_map: Arc<DashMap<String, oneshot::Sender<MessageToClient>>>,
+    event_tx: broadcast::Sender<MessageToClient>,
+    shutdown_tx: watch::Sender<()>,
+}
+
+impl<MessageToServer, MessageToClient> TypedUnixSocketClient<MessageToServer, MessageToClient>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    pub async fn new<P: AsRef<Path>>(socket_path: P) -> Result<Self, Error> {
+        let (tx, _) = broadcast::channel(100);
+        let response_map = Arc::new(DashMap::new());
+        let (event_tx, _) = broadcast::channel(100);
+        let (shutdown_tx, shutdown_rx) = watch::channel(());
+
+        tokio::spawn({
+            let socket_path = socket_path.as_ref().to_path_buf();
+            let tx = tx.clone();
+            let response_map = Arc::clone(&response_map);
+            let event_tx = event_tx.clone();
+            let mut shutdown_rx = shutdown_rx.clone();
+            async move {
+                let mut backoff = get_quick_backoff();
+                loop {
+                    match UnixStream::connect(&socket_path).await {
+                        Ok(stream) => {
+                            handle_connection(
+                                stream,
+                                tx.subscribe(),
+                                Arc::clone(&response_map),
+                                event_tx.clone(),
+                                shutdown_rx.clone(),
+                            )
+                            .await
+                            .unwrap_or_else(|e| {
+                                tracing::error!("Error handling connection: {}", e);
+                            });
+                            break;
+                        }
+                        Err(e) => {
+                            tracing::error!("Error connecting to server: {}", e);
+                            backoff.wait().await;
+                        }
+                    }
+                }
+            }
+        });
+        Ok(Self {
+            tx,
+            response_map,
+            event_tx,
+            shutdown_tx,
+        })
+    }
+
+    pub async fn send_request(&self, request: MessageToServer) -> Result<MessageToClient, Error> {
+        let (response_tx, response_rx) = oneshot::channel();
+        let id = random_token();
+
+        let wrapper = WrappedMessage {
+            id: Some(id.clone()),
+            message: request,
+        };
+
+        self.response_map.insert(id, response_tx);
+
+        // Wait until there is at least one subscriber
+        let mut backoff = get_quick_backoff();
+        while self.tx.receiver_count() == 0 {
+            tracing::info!("Waiting for a subscriber...");
+            backoff.wait().await;
```

Comment:
Another spot where we might want to return an error after some period of time, so this doesn't loop forever.